src/ledger.ts

```diff
@@ -0,0 +1,164 @@
+        for (const bookId of Object.keys(bookGroupedEntries)) {
+            const book = await this.dataConnector.getBook(bookId);
+            if (book && book.restrictions) {
+                const bookEntries = bookGroupedEntries[bookId];
+                const bookBalances = await this.getBookBalances(bookId);
+                const assetGroupedEntries = groupBy(bookEntries, "assetId");
+                Object.keys(assetGroupedEntries).forEach((assetId) => {
+                    const assetBalance = (bookBalances as any)[assetId] || "0";
+                    let newEntriesSum
+                    if (assetGroupedEntries[assetId].length > 1) {
+                        newEntriesSum = assetGroupedEntries[assetId].reduce((r: string|IPostingEntry, i: IPostingEntry) => (new BigNumber(typeof r === "string" ? r : r.value).plus(new BigNumber(i.value))).toString());
+                    } else {
+                        newEntriesSum = assetGroupedEntries[assetId][0].value;
+                    }
+                    if (book.restrictions && book.restrictions.minBalance && (new BigNumber(newEntriesSum).plus(new BigNumber(assetBalance))).isLessThan(new BigNumber(book.restrictions.minBalance))) {
+                        throw new Error(`Minimum credit balance required on book ${bookId} is ${book.restrictions.minBalance} ${assetId} | Current book balance: ${assetBalance} ${assetId}`);
+                    }
+                })
+            }
+        }
+    }
+    public async getOperation(operationId: string) {
+        return this.dataConnector.getOperation(operationId);
+    }
+    private async postOperationEntries(operationId: string, entriesRequests: IPostingEntryRequest[]) {
+        await sleep(2000);   // To test the background queue
+        await this.dataConnector.updateOperationStatus(operationId, OperationStatus.PROCESSING);
+        // TODO: validate the operation
+        // validate the entries using balance restrictions
+        try {
+            await this.validateEntries(entriesRequests);
+        } catch (error) {
+            await this.dataConnector.updateOperationStatus(operationId, OperationStatus.REJECTED, error.message)
+            return;
+        }
+        const entries = entriesRequests.map((entryRequest) => Object.assign({operationId}, entryRequest) as IPostingEntry);
+        await this.dataConnector.insertMultipleEntries(entries);
+        await this.dataConnector.updateOperationStatus(operationId, OperationStatus.APPLIED);
+    }
+    public async postOperation(operationRequest: IOperationRequest, sync?: boolean): Promise<string> {
+        await this.initPromise;
+        const operation: IOperation = Object.assign({status: OperationStatus.INIT}, operationRequest);
+        return new Promise(async (resolve, reject) => {
+            const operationId = (await this.dataConnector.insertOperation(operation)).id as string;
+            this.postingQueue.enqueueTask(async () => {
```

Comment:
Ok so you're doing 2 things here - 

inserting operation in DB
inserting it in an in-memory task queue to be processed.

The latter in fact should be the responsibility of a separate decoupled system which reads the operation from DB and inserts it in an in-memory task queue to be processed.